File: javascript/08_list_models.js
/**
 * 08 - List Available Models
 *
 * This example demonstrates how to retrieve all available models
 * from the Hibana API. This is useful for discovering what models
 * are enabled for your account.
 *
 * Endpoint: GET /v1/models
 */

import OpenAI from 'openai';

// Initialize the Hibana client
const client = new OpenAI({
    apiKey: "YOUR_API_KEY",
    baseURL: "https://api-ai.hibanacloud.com/v1"
});

async function listAllModels() {
    /**
     * List all available models
     */

    console.log("=".repeat(60));
    console.log("Listing All Available Models");
    console.log("=".repeat(60));

    // Get list of models
    const models = await client.models.list();

    console.log(`\nTotal models available: ${models.data.length}\n`);

    // Display each model
    models.data.forEach((model, i) => {
        console.log(`${i + 1}. ${model.id}`);
        console.log(`   Owner: ${model.owned_by}`);
        if (model.created) {
            const createdDate = new Date(model.created * 1000);
            console.log(`   Created: ${createdDate.toISOString().split('T')[0]}`);
        }
        console.log();
    });
}

async function groupModelsByProvider() {
    /**
     * Group models by their provider (OpenAI, Anthropic, etc.)
     */

    console.log("=".repeat(60));
    console.log("Models Grouped by Provider");
    console.log("=".repeat(60));

    const models = await client.models.list();

    // Group models by owner/provider
    const grouped = {};
    models.data.forEach(model => {
        const provider = model.owned_by;
        if (!grouped[provider]) {
            grouped[provider] = [];
        }
        grouped[provider].push(model.id);
    });

    // Display grouped models
    for (const [provider, modelList] of Object.entries(grouped)) {
        console.log(`\n${provider.toUpperCase()}`);
        console.log("-".repeat(40));
        modelList.forEach(modelId => {
            console.log(`  " ${modelId}`);
        });
    }

    console.log(`\nTotal providers: ${Object.keys(grouped).length}`);
}

async function filterModelsByType() {
    /**
     * Filter and categorize models by their type
     */

    console.log("\n" + "=".repeat(60));
    console.log("Models by Type");
    console.log("=".repeat(60));

    const models = await client.models.list();

    // Categorize models
    const chatModels = [];
    const imageModels = [];
    const otherModels = [];

    models.data.forEach(model => {
        const modelId = model.id.toLowerCase();

        if (modelId.includes('dall-e') || modelId.includes('dalle')) {
            imageModels.push(model.id);
        } else if (['gpt', 'claude', 'deepseek', 'gemini'].some(x => modelId.includes(x))) {
            chatModels.push(model.id);
        } else {
            otherModels.push(model.id);
        }
    });

    // Display categorized models
    console.log("\nCHAT/TEXT MODELS:");
    console.log("-".repeat(40));
    chatModels.forEach(model => {
        console.log(`  " ${model}`);
    });

    console.log("\n\nIMAGE GENERATION MODELS:");
    console.log("-".repeat(40));
    imageModels.forEach(model => {
        console.log(`  " ${model}`);
    });

    if (otherModels.length > 0) {
        console.log("\n\nOTHER MODELS:");
        console.log("-".repeat(40));
        otherModels.forEach(model => {
            console.log(`  " ${model}`);
        });
    }
}

async function getSpecificModelInfo() {
    /**
     * Get information about a specific model
     */

    console.log("\n" + "=".repeat(60));
    console.log("Specific Model Information");
    console.log("=".repeat(60));

    const modelId = "gpt-5-nano";

    console.log(`\nQuerying model: ${modelId}`);

    try {
        // Retrieve specific model
        const model = await client.models.retrieve(modelId);

        console.log("\nModel Details:");
        console.log("-".repeat(40));
        console.log(`ID: ${model.id}`);
        console.log(`Object: ${model.object}`);
        console.log(`Owned by: ${model.owned_by}`);

        if (model.created) {
            const createdDate = new Date(model.created * 1000);
            console.log(`Created: ${createdDate.toISOString()}`);
        }

        if (model.permission) {
            console.log(`Permissions: ${model.permission.length} permission(s)`);
        }

    } catch (error) {
        console.log(`Error retrieving model: ${error.message}`);
    }
}

async function checkModelAvailability() {
    /**
     * Check if specific models are available
     */

    console.log("\n" + "=".repeat(60));
    console.log("Model Availability Check");
    console.log("=".repeat(60));

    // Models to check
    const modelsToCheck = [
        "gpt-5-nano",
        "claude-haiku-4-5",
        "deepseek-chat",
        "gemini-2.5-flash-lite",
        "dall-e-3"
    ];

    console.log("\nChecking availability...\n");

    // Get all available models
    const availableModels = await client.models.list();
    const availableIds = availableModels.data.map(m => m.id);

    // Check each model
    modelsToCheck.forEach(modelId => {
        const isAvailable = availableIds.includes(modelId);
        const status = isAvailable ? " Available" : " Not Available";
        console.log(`${status.padEnd(20)} ${modelId}`);
    });
}

async function displayModelsTable() {
    /**
     * Display models in a formatted table
     */

    console.log("\n" + "=".repeat(60));
    console.log("Models Table");
    console.log("=".repeat(60));

    const models = await client.models.list();

    // Table header
    console.log(`\n${"Model ID".padEnd(40)} ${"Provider".padEnd(15)} ${"Type".padEnd(10)}`);
    console.log("-".repeat(70));

    // Table rows
    models.data.forEach(model => {
        const modelId = model.id;
        const provider = model.owned_by;

        // Determine type
        let modelType;
        if (modelId.toLowerCase().includes('dall-e')) {
            modelType = "Image";
        } else if (['gpt', 'claude', 'deepseek', 'gemini'].some(x => modelId.toLowerCase().includes(x))) {
            modelType = "Chat";
        } else {
            modelType = "Other";
        }

        console.log(`${modelId.padEnd(40)} ${provider.padEnd(15)} ${modelType.padEnd(10)}`);
    });

    console.log();
}

// Run the examples
(async () => {
    try {
        // Example 1: List all models
        await listAllModels();

        // Example 2: Group by provider
        await groupModelsByProvider();

        // Example 3: Filter by type
        await filterModelsByType();

        // Example 4: Get specific model info
        await getSpecificModelInfo();

        // Example 5: Check availability
        await checkModelAvailability();

        // Example 6: Display as table
        await displayModelsTable();

    } catch (error) {
        console.error(`\nError: ${error.message}`);
    }
})();

File: python/08_list_models.py
"""
08 - List Available Models

This example demonstrates how to retrieve all available models
from the Hibana API. This is useful for discovering what models
are enabled for your account.

Endpoint: GET /v1/models
"""

from openai import OpenAI
from datetime import datetime

# Initialize the Hibana client
client = OpenAI(
    api_key="YOUR_API_KEY",
    base_url="https://api-ai.hibanacloud.com/v1"
)

def list_all_models():
    """List all available models"""

    print("="*60)
    print("Listing All Available Models")
    print("="*60)

    # Get list of models
    models = client.models.list()

    print(f"\nTotal models available: {len(models.data)}\n")

    # Display each model
    for i, model in enumerate(models.data, 1):
        print(f"{i}. {model.id}")
        print(f"   Owner: {model.owned_by}")
        if hasattr(model, 'created'):
            created_date = datetime.fromtimestamp(model.created)
            print(f"   Created: {created_date.strftime('%Y-%m-%d')}")
        print()

def group_models_by_provider():
    """Group models by their provider (OpenAI, Anthropic, etc.)"""

    print("="*60)
    print("Models Grouped by Provider")
    print("="*60)

    models = client.models.list()

    # Group models by owner/provider
    grouped = {}
    for model in models.data:
        provider = model.owned_by
        if provider not in grouped:
            grouped[provider] = []
        grouped[provider].append(model.id)

    # Display grouped models
    for provider, model_list in grouped.items():
        print(f"\n{provider.upper()}")
        print("-" * 40)
        for model_id in model_list:
            print(f"  {model_id}")

    print(f"\nTotal providers: {len(grouped)}")

def filter_models_by_type():
    """Filter and categorize models by their type"""

    print("\n" + "="*60)
    print("Models by Type")
    print("="*60)

    models = client.models.list()

    # Categorize models
    chat_models = []
    image_models = []
    other_models = []

    for model in models.data:
        model_id = model.id.lower()

        if 'dall-e' in model_id or 'dalle' in model_id:
            image_models.append(model.id)
        elif any(x in model_id for x in ['gpt', 'claude', 'deepseek', 'gemini']):
            chat_models.append(model.id)
        else:
            other_models.append(model.id)

    # Display categorized models
    print("\nCHAT/TEXT MODELS:")
    print("-" * 40)
    for model in chat_models:
        print(f"   {model}")

    print("\n\nIMAGE GENERATION MODELS:")
    print("-" * 40)
    for model in image_models:
        print(f"   {model}")

    if other_models:
        print("\n\nOTHER MODELS:")
        print("-" * 40)
        for model in other_models:
            print(f"   {model}")

def get_specific_model_info():
    """Get information about a specific model"""

    print("\n" + "="*60)
    print("Specific Model Information")
    print("="*60)

    model_id = "gpt-5-nano"

    print(f"\nQuerying model: {model_id}")

    try:
        # Retrieve specific model
        model = client.models.retrieve(model_id)

        print("\nModel Details:")
        print("-" * 40)
        print(f"ID: {model.id}")
        print(f"Object: {model.object}")
        print(f"Owned by: {model.owned_by}")

        if hasattr(model, 'created'):
            created_date = datetime.fromtimestamp(model.created)
            print(f"Created: {created_date.strftime('%Y-%m-%d %H:%M:%S')}")

        if hasattr(model, 'permission'):
            print(f"Permissions: {len(model.permission)} permission(s)")

    except Exception as e:
        print(f"Error retrieving model: {e}")

def check_model_availability():
    """Check if specific models are available"""

    print("\n" + "="*60)
    print("Model Availability Check")
    print("="*60)

    # Models to check
    models_to_check = [
        "gpt-5-nano",
        "claude-haiku-4-5",
        "deepseek-chat",
        "gemini-2.5-flash-lite",
        "dall-e-3"
    ]

    print("\nChecking availability...\n")

    # Get all available models
    available_models = client.models.list()
    available_ids = [m.id for m in available_models.data]

    # Check each model
    for model_id in models_to_check:
        is_available = model_id in available_ids
        status = " Available" if is_available else " Not Available"
        print(f"{status.ljust(20)} {model_id}")

def display_models_table():
    """Display models in a formatted table"""

    print("\n" + "="*60)
    print("Models Table")
    print("="*60)

    models = client.models.list()

    # Table header
    print(f"\n{'Model ID':<40} {'Provider':<15} {'Type':<10}")
    print("-" * 70)

    # Table rows
    for model in models.data:
        model_id = model.id
        provider = model.owned_by

        # Determine type
        if 'dall-e' in model_id.lower():
            model_type = "Image"
        elif any(x in model_id.lower() for x in ['gpt', 'claude', 'deepseek', 'gemini']):
            model_type = "Chat"
        else:
            model_type = "Other"

        print(f"{model_id:<40} {provider:<15} {model_type:<10}")

    print("\n")

if __name__ == "__main__":
    try:
        # Example 1: List all models
        list_all_models()

        # Example 2: Group by provider
        group_models_by_provider()

        # Example 3: Filter by type
        filter_models_by_type()

        # Example 4: Get specific model info
        get_specific_model_info()

        # Example 5: Check availability
        check_model_availability()

        # Example 6: Display as table
        display_models_table()

    except Exception as e:
        print(f"\nError: {e}")
